README.md
Maximilian3 | Script Runner
==========

Redshift does not have stored procedures so we will use a simple script runner to execute our ELT and load tasks.

Both script runner and SQLWorkbenchJ will use the same notation for variable expansion: `$[?variable]`.  This string will be replaced at runtime by the appropriate variable:

There are three types of substitution parameters that can be used through script runner.  The first is a set of standard etl params for ease of use:

* -f, from_date, default 1776-07-24
* -t,to_date, default 9999-12-31
* -b, batch_no, default is -1

The second option is for arbitrary variable expansion.  This is passed in the following format because rundeck commands don't like json quotes:

* -p, example: -p "param1-val1, param2-val2"

The final option is to use the batchy integration

* -wf, batchy_job, this will substitute parameters from a batchy workflow, this should be a fully qualified batchy job name of the format wf.job, if no job is specified it will assume global


Here is a sample SQL Script.  If run in SQL workbench you will be prompted for values for var1 and var2.

```
drop table if exists  zzztemp;

create table zzztemp (
  dt timestamp,
  var varchar,
  from_date timestamp,
  batch_no integer
  );

insert into zzztemp
values (getdate(), '$[?var1]', '$[?from_date]', '$[?batch_no]');

insert into zzztemp
values (getdate(), '$[?var2]', '$[?from_date]', '$[?batch_no]');

select * from zzztemp;

```

In the script runner you would use the following params to substitute that value.  It is assumed these params would be dynamically substituted by the calling script or informatica process:

`python -m script_runner.module -s "sample/script_runner_test.sql" -p "var1-cat, var2-dog" -f '1980-12-31 07:00' -b '9999'`

Assuming you had workflow config in batchy under wf3, you could also use this script:

`python -m  script_runner.module -s sample/script_runner_test.sql -wf wf3`


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added

## [0.1.4] - 2020-11-03
### Changed
    - Fix issue when parsing batchy work flow
    - Handle batchy connection issues

## [0.1.3] - 2020-03-19
### Changed
    - Secret manager default project name

## [0.1.2] - 2020-03-18
### Changed
    - Added documentation how to use config manager
    - Secret manager config parameter
    - Fix issue when getting Batch config

## [0.1.1] - 2020-03-17
### Added
    - Add unit tests
    - Add default from date
### Changed
    - Modify scripty to make it executable as script and as module call

## [0.1.0] - 2020-03-11
### Changed
    - Initial project setup
    - Config Management
    - Batch Control

[0.1.4]: https://github.com/equinoxfitness/scripty/compare/v0.1.3...v0.1.4
[0.1.3]: https://github.com/equinoxfitness/scripty/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/equinoxfitness/scripty/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/equinoxfitness/scripty/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/equinoxfitness/scripty/releases/tag/v0.1.0

tests/test_scripty.py
import unittest
from scripty.module import ScriptRunner
from unittest.mock import patch


class TestScriptRunner(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.scripty = ScriptRunner(
            db_name="db_name",
            host="host",
            user="user",
            password="password",
            port=5439,
            batchy_server="batchy_server",
            batchy_port="1111",
            test=True,
        )
        cls.sql = """
                SELECT field1, field2 FROM test
                WHERE field1 = '$[?var1]' AND field2 = '$[?var2]'
                """

    @patch("scripty.module.ScriptRunner.expand_params")
    def test_expand_params(self, mock_expand_params):
        print("--------------test_expand_params")

        mock_expand_params.return_value = """
                    SELECT field1, field2 FROM test
                    WHERE field1 = 'result1' AND field2 = 'result2'
                    """

        params = {"var1": "result1", "var2": "result2"}
        result = self.scripty.expand_params(self.sql, params)
        expected_result = """
                    SELECT field1, field2 FROM test
                    WHERE field1 = 'result1' AND field2 = 'result2'
                    """

        self.assertEqual(result, expected_result)

    @patch("scripty.module.ScriptRunner.run_script")
    def test_run_script(self, mock_run_script):
        print("--------------test_run_script")
        self.scripty.run_script(self.sql)
        self.assertTrue(mock_run_script.called)


if __name__ == "__main__":
    unittest.main()


scripty/config_wrapper.py
from datacoco_core.config import config
from datacoco_secretsmanager import SecretsManager


class ConfigWrapper:
    """
    Wrapper file for config management for script_runner.
    """

    @staticmethod
    def sm_conf(project_name:str, team_name:str):
        """
        Simple config wrapper for using secrets manager.
        """
        c = SecretsManager().get_config(project_name, team_name)
        return c

    @staticmethod
    def extend_parser(parser):
        parser.add_argument(
            "-c",
            "--config",
            default="core",
            help="""
                whether to use secret_manager or
                datacoco_core to retrieve secrets
                """,
            choices=["secret_manager", "core"],
        )

        parser.add_argument(
            "-smp",
            "--secret_project_name",
            default="Maximilian3",
            help="""
                secret manager project group
                """
        )

        parser.add_argument(
            "-smt",
            "--secret_team",
            default="data",
            help="""
                secret manager team
                """
        )

        return parser

    @staticmethod
    def process_config(args):
        if args.config == "secret_manager":
            conf = ConfigWrapper.sm_conf(project_name=args.secret_project_name, team_name=args.secret_team)
        elif args.config == "core":
            conf = config()

        db_name = conf[args.database]["db_name"]
        host = conf[args.database]["host"]
        user = conf[args.database]["user"]
        password = conf[args.database]["password"]
        port = conf[args.database]["port"]
        batchy_server = conf["batchy"]["server"]
        batchy_port = conf["batchy"]["port"]

        return db_name, host, user, password, port, batchy_server, batchy_port


scripty/module.py
import sys
import os
import argparse
from datacoco_core.logger import Logger
from datacoco_batch import Batch
from datacoco_db.pg_tools import PGInteraction
from pathlib import Path
from scripty.config_wrapper import ConfigWrapper
from scripty import DEFAULT_FROM_DATE


class ScriptRunner:
    """
    Generic class for execution of a parameterized script
    in postgres or redshift.
    """

    def __init__(self, db_name, host, user, password, port, batchy_server, batchy_port, test=False):
        self.pg = PGInteraction(
            dbname=db_name,
            host=host,
            user=user,
            password=password,
            port=port,
            schema="public",
        )

        self.batchy_server = batchy_server
        self.batchy_port = batchy_port

        if not test:
            self.pg.conn()

    @staticmethod
    def expand_params(sql, params):
        """
        substitutes params in sql stagement
        :param sql:
        :param params:
        :return: sql, expanded with params
        """
        for p in params.keys():
            var = "$[?" + p + "]"
            val = str(params[p])
            sql = sql.replace(var, val)
        return sql

    def run_script(
        self,
        script,
        from_date=None,
        to_date=None,
        batch_id=None,
        params=None,
        batchy_job=None,
    ):
        """
        Method for expanding and running sql statements
        :param script:
        :param from_date:
        :param to_date:
        :param batch_id:
        :param params:
        :return:
        """

        paramset = {}

        # set up logger
        script_path, script_filename = os.path.split(script)
        logger = Logger(logname=script_filename)
        sys.excepthook = logger.handle_exception

        # Check if file exists
        file = Path(f"{script_path}/{script_filename}").is_file()
        if not file:
            e = "File not found, please check path"
            logger.l(e)
            raise RuntimeError(e)

        # first we retrive params  we will load these into dict first,
        # any additional params specified will override
        if batchy_job:
            try:
                batchy_params = Batch(
                    batchy_job,
                    server=self.batchy_server,
                    port=self.batchy_port,
                ).get_status()
            except Exception as e:
                logger.l(str(e))
                raise Exception(str(e))

            if batchy_params is None:
                error = "Unable to connect to batchy"
                logger.l(error)
                raise Exception(error)

            if 'error' in batchy_params:
                error = f"Batchy encountered an error: {batchy_params['message']}"
                logger.l(error)
                raise Exception(error)

            paramset.update(batchy_params['global'])

        # next we apply custom params and special metadata fields,
        # again this will overrite batchy params if specified
        # convert string params to dict
        try:
            params = dict(
                (k.strip(), v.strip())
                for k, v in (item.split("-") for item in params.split(","))
            )
        except Exception as e:
            logger.l(f"issue parsing params: {e}")

        if isinstance(params, dict):
            paramset.update(params)

        if from_date:
            paramset["from_date"] = from_date
        if to_date:
            paramset["to_date"] = to_date
        if batch_id:
            paramset["batch_id"] = batch_id

        print(paramset)

        # now defaults for special metadata fields
        if paramset.get("from_date") is None:
            paramset["from_date"] = DEFAULT_FROM_DATE
        if paramset.get("to_date") is None:
            paramset["to_date"] = "9999-12-31"
        if paramset.get("batch_id") is None:
            paramset["batch_id"] = "-1"
        # we'll keep batch_no for backwards compatibility
        paramset["batch_no"] = paramset["batch_id"]

        raw_sql = open(script).read()
        sql = self.expand_params(raw_sql, paramset)
        sql_message = (
            "\n\n--sql script start:\n" + sql + "\n--sql script end\n\n"
        )
        logger.l(sql_message)

        self.pg.batch_open()

        logger.l("starting script")
        try:
            self.pg.exec_sql(sql)
            self.pg.batch_commit()
            logger.l("batch commit")
        except Exception as e:
            logger.l("execution failed with error: " + str(e))
            raise RuntimeError(e)
            # os._exit(1) #this truncates logout in rundeck

def main(args):
    # argparse
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-s", "--script", help="""enter a path to your script """
    )
    parser.add_argument(
        "-p",
        "--parameters",  # type=json.loads, default='{"none":"none"}',
        default="none-none",
        help="""
            additional params to be substituted in script,
            example: -p param1-val1, param2-val2
            """,
    )
    parser.add_argument(
        "-d",
        "--database",
        help="""db alias from etl.cfg, default is cosmo """,
        default="cosmo",
    )

    parser.add_argument(
        "-f", "--from_date", help="""from_date""", default=None
    )
    parser.add_argument("-t", "--to_date", help="""to_date""", default=None)
    parser.add_argument(
        "-b", "--batch_id", help="""enter batch id """, default=None
    )
    parser.add_argument(
        "-wf",
        "--batchy_job",
        help="""fully qualified batchy job name of the format wf.job""",
        default=None,
    )

    parser = ConfigWrapper.extend_parser(parser)

    args = parser.parse_args(args)

    (db_name, host, user, password, port, batchy_server, batchy_port) = ConfigWrapper.process_config(args)

    ScriptRunner(db_name, host, user, password, port, batchy_server, batchy_port).run_script(
        args.script,
        args.from_date,
        args.to_date,
        args.batch_id,
        args.parameters,
        args.batchy_job,
    )

if __name__ == "__main__":
    main(sys.argv[1:])
